### PR TITLE
ISPN-14144 Propagate subject in task execution

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-14144

Properly fix this

### CODE REVIEW

File: server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTask.java [hunk 1]
```diff
@@ -30,14 +26,14 @@
    @ProtoField(2)
    final String cacheName;
 
-   @ProtoField(number = 3, collectionImplementation = ArrayList.class)
-   final List<TaskParameter> parameters;
+   @ProtoField(3)
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTask.java:29]: The correct way to handle this is to use a new field. Ideally we would just be able to add `@ProtoReservice(value=3)` to the class and `proto.lock` checks would pass. However, the `reserved` statement is currently commented due to a limitation in the protostream proto parser https://github.com/infinispan/protostream/issues/174.

Adding `@ProtoReservice(value=3)` will still prevent the field being used in the `ProtoField` though.

A workaround to keep proto compatibility would be todo the following:

```java
   @ProtoField(1)
   final String taskName;

   @ProtoField(2)
   final String cacheName;

   @ProtoField(number = 3, collectionImplementation = ArrayList.class)
   List<TaskParameter> parameters;

   @ProtoField(4)
   final TaskContext context;

   @ProtoFactory
   DistributedServerTask(String taskName, String cacheName, List<TaskParameter> parameters, TaskContext context) {
      this(taskName, cacheName, context);
   }
   
   public DistributedServerTask(String taskName, String cacheName, TaskContext context) {
      this.cacheName = cacheName;
      this.taskName = taskName;
      this.context = context;
   }
```

**Review by ryanemerson — CHANGES_REQUESTED** [review]

File: tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java [hunk 2]
```diff
@@ -12,23 +13,27 @@
...
  * @since 8.1
  */
 public class TaskContext {
-   private EmbeddedCacheManager cacheManager;
-   private Optional<Marshaller> marshaller = Optional.empty();
-   private Optional<Cache<?, ?>> cache = Optional.empty();
-   private Optional<Map<String, ?>> parameters = Optional.empty();
-   private Optional<Subject> subject = Optional.empty();
-   private boolean logEvent;
+   private transient EmbeddedCacheManager cacheManager;
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java:22]: I know, but it is a nice way to mark the fields

File: server/runtime/src/main/java/org/infinispan/server/tasks/TaskContextAdapter.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.infinispan.protostream.annotations.ProtoField;
+import org.infinispan.protostream.annotations.ProtoName;
+import org.infinispan.tasks.TaskContext;
+
+/**
+ * @since 14.0
+ **/
+@ProtoAdapter(TaskContext.class)
+@ProtoName("TaskContext")
+public class TaskContextAdapter {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/tasks/TaskContextAdapter.java]: Done

File: tasks/api/src/main/java/org/infinispan/tasks/Task.java [hunk 4]
```diff
@@ -32,6 +32,15 @@ default TaskExecutionMode getExecutionMode() {
       return TaskExecutionMode.ONE_NODE;
    }
 
+   /**
+    * Whether tasks should reuse a single instance or create a new instance per execution. {@link TaskInstantiationMode#SHARED} is the default
+    *
+    * @return TaskInstantiationMode
+    */
+   default TaskInstantiationMode getInstantiationMode() {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply tasks/api/src/main/java/org/infinispan/tasks/Task.java:40]: Done

File: tasks/api/src/main/java/org/infinispan/tasks/TaskInstantiationMode.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+public enum TaskInstantiationMode {
+   /**
+    * Creates a single instance that is reused for every task execution
+    */
+   SHARED,
+   /**
+    * Creates a new instance for every invocation
+    */
+   ISOLATED
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply tasks/api/src/main/java/org/infinispan/tasks/TaskInstantiationMode.java:11]: I have created a test for each mode.

File: server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTask.java (hunk 1 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/tasks/DistributedServerTask.java:29]: This is for internal invocation only, so I'm not fussed about losing backwards compat.

**Comment by ryanemerson** [comment]
`ScriptingTest` and `ScriptingDataStoresTest` are both failing.

**Comment by tristantarrant** [comment]
Should be fixed. The `AuthorizationCertIT` failures made me discover that we were adding an `anonymous` principal to the subject when using client cert authentication. I've fixed that.

**Pull request merged**